File: sdk/js/src/implementations/AdminMethods.ts
import type { SharpAISdk } from '../SharpAISdk';
import type {
  CreateAssignmentRequest,
  CreateUserRequest,
  CreatedCredential,
  IAdminMethods,
  LoginResponse,
  QueryParams,
} from '../interfaces/IAdminMethods';

/**
 * Implementation of administrative methods: settings, request history, authentication, audit, and
 * account/RBAC management.
 */
export class AdminMethods implements IAdminMethods {
  private readonly sdk: SharpAISdk;

  constructor(sdk: SharpAISdk) {
    this.sdk = sdk;
  }

  private url(path: string, query?: QueryParams): string {
    let url = `${this.sdk.endpoint}${path}`;
    if (query) {
      const params = new URLSearchParams();
      for (const [key, value] of Object.entries(query)) {
        if (value !== undefined && value !== null && value !== '') params.append(key, String(value));
      }
      const qs = params.toString();
      if (qs) url = `${url}?${qs}`;
    }
    return url;
  }

  // ---- settings ----

  getSettings(signal?: AbortSignal): Promise<unknown | null> {
    return this.sdk.sendAsync('GET', this.url('/api/settings'), undefined, signal);
  }

  updateSettings(settings: unknown, signal?: AbortSignal): Promise<unknown | null> {
    return this.sdk.sendAsync('PUT', this.url('/api/settings'), { body: settings }, signal);
  }

  // ---- request history ----

  requestHistory(query?: QueryParams, signal?: AbortSignal): Promise<unknown | null> {
    return this.sdk.sendAsync('GET', this.url('/v1.0/api/request-history', query), undefined, signal);
  }

  requestHistorySummary(query?: QueryParams, signal?: AbortSignal): Promise<unknown | null> {
    return this.sdk.sendAsync('GET', this.url('/v1.0/api/request-history/summary', query), undefined, signal);
  }

  requestHistoryEntry(id: string, signal?: AbortSignal): Promise<unknown | null> {
    return this.sdk.sendAsync('GET', this.url(`/v1.0/api/request-history/${encodeURIComponent(id)}`), undefined, signal);
  }

  // ---- authentication ----

  async login(email: string, password: string, tenantGuid?: string, signal?: AbortSignal): Promise<LoginResponse | null> {
    const headers: Record<string, string> = { 'x-email': email, 'x-password': password };
    if (tenantGuid) headers['x-tenant-guid'] = tenantGuid;
    const result = await this.sdk.sendAsync<LoginResponse>('POST', this.url('/v1.0/token'), { headers }, signal);
    if (result?.token) this.sdk.token = result.token;
    return result;
  }

  session(signal?: AbortSignal): Promise<unknown | null> {
    return this.sdk.sendAsync('GET', this.url('/v1.0/token'), undefined, signal);
  }

  async logout(signal?: AbortSignal): Promise<unknown | null> {
    const result = await this.sdk.sendAsync('DELETE', this.url('/v1.0/token'), undefined, signal);
    this.sdk.token = null;
    return result;
  }

  audit(query?: QueryParams, signal?: AbortSignal): Promise<unknown | null> {
    return this.sdk.sendAsync('GET', this.url('/v1.0/api/audit', query), undefined, signal);
  }

  // ---- account / RBAC management ----

  listTenants(query?: QueryParams, signal?: AbortSignal): Promise<unknown | null> {
    return this.sdk.sendAsync('GET', this.url('/v1.0/tenants', query), undefined, signal);
  }

  createTenant(name: string, signal?: AbortSignal): Promise<unknown | null> {
    return this.sdk.sendAsync('POST', this.url('/v1.0/tenants'), { body: { Name: name } }, signal);
  }

  listUsers(tenantGuid: string, query?: QueryParams, signal?: AbortSignal): Promise<unknown | null> {
    return this.sdk.sendAsync('GET', this.url(`/v1.0/tenants/${tenantGuid}/users`, query), undefined, signal);
  }

  createUser(tenantGuid: string, request: CreateUserRequest, signal?: AbortSignal): Promise<unknown | null> {
    return this.sdk.sendAsync('POST', this.url(`/v1.0/tenants/${tenantGuid}/users`), { body: request }, signal);
  }

  deleteUser(tenantGuid: string, userGuid: string, signal?: AbortSignal): Promise<unknown | null> {
    return this.sdk.sendAsync('DELETE', this.url(`/v1.0/tenants/${tenantGuid}/users/${userGuid}`), undefined, signal);
  }

  listCredentials(tenantGuid: string, query?: QueryParams, signal?: AbortSignal): Promise<unknown | null> {
    return this.sdk.sendAsync('GET', this.url(`/v1.0/tenants/${tenantGuid}/credentials`, query), undefined, signal);
  }

  createCredential(tenantGuid: string, userGuid: string, name: string, signal?: AbortSignal): Promise<CreatedCredential | null> {
    return this.sdk.sendAsync<CreatedCredential>(
      'POST',
      this.url(`/v1.0/tenants/${tenantGuid}/credentials`),
      { body: { UserGuid: userGuid, Name: name } },
      signal
    );
  }

  listRoles(tenantGuid: string, query?: QueryParams, signal?: AbortSignal): Promise<unknown | null> {
    return this.sdk.sendAsync('GET', this.url(`/v1.0/tenants/${tenantGuid}/roles`, query), undefined, signal);
  }

  createAssignment(tenantGuid: string, request: CreateAssignmentRequest, signal?: AbortSignal): Promise<unknown | null> {
    return this.sdk.sendAsync('POST', this.url(`/v1.0/tenants/${tenantGuid}/assignments`), { body: request }, signal);
  }

  userPermissions(tenantGuid: string, userGuid: string, signal?: AbortSignal): Promise<unknown | null> {
    return this.sdk.sendAsync('GET', this.url(`/v1.0/tenants/${tenantGuid}/users/${userGuid}/permissions`), undefined, signal);
  }

  credentialPermissions(tenantGuid: string, credentialGuid: string, signal?: AbortSignal): Promise<unknown | null> {
    return this.sdk.sendAsync('GET', this.url(`/v1.0/tenants/${tenantGuid}/credentials/${credentialGuid}/permissions`), undefined, signal);
  }
}

File: sdk/csharp/src/SharpAI.Sdk/Implementations/AdminMethods.cs
namespace SharpAI.Sdk.Implementations
{
    using System;
    using System.Collections.Generic;
    using System.Net.Http;
    using System.Text.Json;
    using System.Threading;
    using System.Threading.Tasks;

    using SharpAI.Sdk.Interfaces;
    using SharpAI.Sdk.Models.Admin;

    /// <summary>
    /// Implementation of administrative methods: settings, request history, authentication, audit, and
    /// account/RBAC management.
    /// </summary>
    public class AdminMethods : IAdminMethods
    {
        private readonly SharpAISdk _Sdk;

        /// <summary>
        /// Initialize the administrative methods.
        /// </summary>
        /// <param name="sdk">Owning SDK instance.</param>
        public AdminMethods(SharpAISdk sdk)
        {
            _Sdk = sdk ?? throw new ArgumentNullException(nameof(sdk));
        }

        private string Url(string path, IDictionary<string, string>? query = null)
        {
            string url = _Sdk.Endpoint + path;
            if (query != null)
            {
                List<string> parts = new List<string>();
                foreach (KeyValuePair<string, string> kv in query)
                {
                    if (!string.IsNullOrEmpty(kv.Value))
                        parts.Add(Uri.EscapeDataString(kv.Key) + "=" + Uri.EscapeDataString(kv.Value));
                }
                if (parts.Count > 0) url += "?" + string.Join("&", parts);
            }
            return url;
        }

        /// <inheritdoc />
        public Task<JsonElement?> GetSettingsAsync(CancellationToken cancellationToken = default)
            => _Sdk.SendAsync<JsonElement?>(HttpMethod.Get, Url("/api/settings"), null, null, cancellationToken);

        /// <inheritdoc />
        public Task<JsonElement?> UpdateSettingsAsync(object settings, CancellationToken cancellationToken = default)
            => _Sdk.SendAsync<JsonElement?>(HttpMethod.Put, Url("/api/settings"), settings, null, cancellationToken);

        /// <inheritdoc />
        public Task<JsonElement?> RequestHistoryAsync(IDictionary<string, string>? query = null, CancellationToken cancellationToken = default)
            => _Sdk.SendAsync<JsonElement?>(HttpMethod.Get, Url("/v1.0/api/request-history", query), null, null, cancellationToken);

        /// <inheritdoc />
        public Task<JsonElement?> RequestHistorySummaryAsync(IDictionary<string, string>? query = null, CancellationToken cancellationToken = default)
            => _Sdk.SendAsync<JsonElement?>(HttpMethod.Get, Url("/v1.0/api/request-history/summary", query), null, null, cancellationToken);

        /// <inheritdoc />
        public Task<JsonElement?> RequestHistoryEntryAsync(string id, CancellationToken cancellationToken = default)
            => _Sdk.SendAsync<JsonElement?>(HttpMethod.Get, Url("/v1.0/api/request-history/" + Uri.EscapeDataString(id)), null, null, cancellationToken);

        /// <inheritdoc />
        public async Task<LoginResponse?> LoginAsync(string email, string password, string? tenantGuid = null, CancellationToken cancellationToken = default)
        {
            Dictionary<string, string> headers = new Dictionary<string, string> { { "x-email", email }, { "x-password", password } };
            if (!string.IsNullOrEmpty(tenantGuid)) headers["x-tenant-guid"] = tenantGuid!;
            LoginResponse? result = await _Sdk.SendAsync<LoginResponse>(HttpMethod.Post, Url("/v1.0/token"), null, headers, cancellationToken).ConfigureAwait(false);
            if (!string.IsNullOrEmpty(result?.Token)) _Sdk.Token = result!.Token;
            return result;
        }

        /// <inheritdoc />
        public Task<JsonElement?> SessionAsync(CancellationToken cancellationToken = default)
            => _Sdk.SendAsync<JsonElement?>(HttpMethod.Get, Url("/v1.0/token"), null, null, cancellationToken);

        /// <inheritdoc />
        public async Task<JsonElement?> LogoutAsync(CancellationToken cancellationToken = default)
        {
            JsonElement? result = await _Sdk.SendAsync<JsonElement?>(HttpMethod.Delete, Url("/v1.0/token"), null, null, cancellationToken).ConfigureAwait(false);
            _Sdk.Token = null;
            return result;
        }

        /// <inheritdoc />
        public Task<JsonElement?> AuditAsync(IDictionary<string, string>? query = null, CancellationToken cancellationToken = default)
            => _Sdk.SendAsync<JsonElement?>(HttpMethod.Get, Url("/v1.0/api/audit", query), null, null, cancellationToken);

        /// <inheritdoc />
        public Task<JsonElement?> ListTenantsAsync(IDictionary<string, string>? query = null, CancellationToken cancellationToken = default)
            => _Sdk.SendAsync<JsonElement?>(HttpMethod.Get, Url("/v1.0/tenants", query), null, null, cancellationToken);

        /// <inheritdoc />
        public Task<JsonElement?> CreateTenantAsync(string name, CancellationToken cancellationToken = default)
            => _Sdk.SendAsync<JsonElement?>(HttpMethod.Post, Url("/v1.0/tenants"), new { Name = name }, null, cancellationToken);

        /// <inheritdoc />
        public Task<JsonElement?> ListUsersAsync(string tenantGuid, IDictionary<string, string>? query = null, CancellationToken cancellationToken = default)
            => _Sdk.SendAsync<JsonElement?>(HttpMethod.Get, Url($"/v1.0/tenants/{tenantGuid}/users", query), null, null, cancellationToken);

        /// <inheritdoc />
        public Task<JsonElement?> CreateUserAsync(string tenantGuid, CreateUserRequest request, CancellationToken cancellationToken = default)
            => _Sdk.SendAsync<JsonElement?>(HttpMethod.Post, Url($"/v1.0/tenants/{tenantGuid}/users"), request, null, cancellationToken);

        /// <inheritdoc />
        public Task<JsonElement?> DeleteUserAsync(string tenantGuid, string userGuid, CancellationToken cancellationToken = default)
            => _Sdk.SendAsync<JsonElement?>(HttpMethod.Delete, Url($"/v1.0/tenants/{tenantGuid}/users/{userGuid}"), null, null, cancellationToken);

        /// <inheritdoc />
        public Task<JsonElement?> ListCredentialsAsync(string tenantGuid, IDictionary<string, string>? query = null, CancellationToken cancellationToken = default)
            => _Sdk.SendAsync<JsonElement?>(HttpMethod.Get, Url($"/v1.0/tenants/{tenantGuid}/credentials", query), null, null, cancellationToken);

        /// <inheritdoc />
        public Task<CreatedCredential?> CreateCredentialAsync(string tenantGuid, string userGuid, string name, CancellationToken cancellationToken = default)
            => _Sdk.SendAsync<CreatedCredential>(HttpMethod.Post, Url($"/v1.0/tenants/{tenantGuid}/credentials"), new { UserGuid = userGuid, Name = name }, null, cancellationToken);

        /// <inheritdoc />
        public Task<JsonElement?> ListRolesAsync(string tenantGuid, IDictionary<string, string>? query = null, CancellationToken cancellationToken = default)
            => _Sdk.SendAsync<JsonElement?>(HttpMethod.Get, Url($"/v1.0/tenants/{tenantGuid}/roles", query), null, null, cancellationToken);

        /// <inheritdoc />
        public Task<JsonElement?> CreateAssignmentAsync(string tenantGuid, CreateAssignmentRequest request, CancellationToken cancellationToken = default)
            => _Sdk.SendAsync<JsonElement?>(HttpMethod.Post, Url($"/v1.0/tenants/{tenantGuid}/assignments"), request, null, cancellationToken);

        /// <inheritdoc />
        public Task<JsonElement?> UserPermissionsAsync(string tenantGuid, string userGuid, CancellationToken cancellationToken = default)
            => _Sdk.SendAsync<JsonElement?>(HttpMethod.Get, Url($"/v1.0/tenants/{tenantGuid}/users/{userGuid}/permissions"), null, null, cancellationToken);

        /// <inheritdoc />
        public Task<JsonElement?> CredentialPermissionsAsync(string tenantGuid, string credentialGuid, CancellationToken cancellationToken = default)
            => _Sdk.SendAsync<JsonElement?>(HttpMethod.Get, Url($"/v1.0/tenants/{tenantGuid}/credentials/{credentialGuid}/permissions"), null, null, cancellationToken);
    }
}
